### PR TITLE
Refactor groth16, ric0 and zksync pallet to pallet-verifiers abstraction

### PR DESCRIPTION
**Pull request by la10736**
Refactor groth16, ric0 and zksync pallet to pallet-verifiers abstraction to have a common interface and behaviour.

- risc0 can bypass the ecdsa hash of its vk and zksync 
- zksync can just use void `()` vk

All pallets compute the statement like

keccak(id)+hash(vk)+keccak(pubs)

where `hash(vk)` is a developer choice where the default is `keccak(VK)`

### CODE REVIEW

File: verifiers/risc0/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    fn pubs_bytes(pubs: &Self::Pubs) -> hp_verifiers::Cow<[u8]> {
+        hp_verifiers::Cow::Borrowed(pubs)
+    }
+
+    fn vk_hash(vk: &Self::Vk) -> H256 {
+        *vk
+    }
+
+    fn vk_bytes(_vk: &Self::Vk) -> hp_verifiers::Cow<[u8]> {
+        panic!("We cannot know vk preimage: use vk_hash() instead")
```
Comment (DanieleDiBenedetto) [inline verifiers/risc0/src/lib.rs:90]: Would leave here just a comment stating that Risc0 vk is already hashed

File: verifiers/zksync/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    ) -> Result<(), VerifyError> {
+        native::zksync_verify::verify(proof, *pubs).map_err(Into::into)
+    }
+
+    fn pubs_bytes(pubs: &Self::Pubs) -> Cow<[u8]> {
+        Cow::Borrowed(pubs)
+    }
+
+    fn vk_bytes(_vk: &Self::Vk) -> Cow<[u8]> {
+        static VOID: &[u8] = &[];
```
Comment (DanieleDiBenedetto) [inline verifiers/zksync/src/lib.rs:58]: Please small comment also here about the rationale

**Review by DanieleDiBenedetto — APPROVED** [review]
Minor comments, looks good for all the rest :)

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -541,6 +517,54 @@ impl pallet_verifiers::Config<pallet_fflonk_verifier::Fflonk> for Runtime {
...
+
+impl pallet_groth16_verifier::Config for Runtime {
+    type MaxNumInputs = Groth16MaxNumInputs;
+}
+
+// We should be sure that the max number of inputs does not exceed the max number of inputs in the verifier crate.
+const_assert!(
+    <Runtime as pallet_groth16_verifier::Config>::MaxNumInputs::get()
+        <= pallet_groth16_verifier::MAX_NUM_INPUTS
+);
```
[the remark this answers is not in the document]
Comment (la10736) [reply runtime/src/lib.rs:540]: To be sure (at compile time) that you don't use something that's greater that the maximum defined in the pallet and used to compute the MaxEncodedLength for the VK

File: verifiers/groth16/src/data_structures.rs [hunk 4]
```diff
@@ -118,6 +138,24 @@ impl<E: Pairing> TryInto<ark_groth16::VerifyingKey<E>> for VerificationKey {
     }
 }
 
+// impl<E: Pairing> TryInto<ark_groth16::VerifyingKey<E>> for VerificationKey {
```
[the remark this answers is not in the document]
Comment (la10736) [reply verifiers/groth16/src/data_structures.rs:141]: THX

File: verifiers/risc0/src/lib.rs (hunk 1 above)
Comment (la10736) [reply verifiers/risc0/src/lib.rs:90]: Ok

File: verifiers/zksync/src/lib.rs (hunk 2 above)
Comment (la10736) [reply verifiers/zksync/src/lib.rs:58]: Ok

File: verifiers/groth16/src/groth16.rs [hunk 5]
```diff
@@ -40,6 +41,18 @@ pub struct VerificationKeyWithCurve {
...
 
+impl MaxEncodedLen for VerificationKeyWithCurve {
+    fn max_encoded_len() -> usize {
+        let g1_size: u32 = G1::max_encoded_len()
+            .try_into()
+            .expect("Should be a valid u32. qed");
+        Curve::max_encoded_len()
+            + G1::max_encoded_len()
+            + 3 * G2::max_encoded_len()
+            + vec_max_encoded_len(crate::MAX_NUM_INPUTS * g1_size)
```
Comment (lgiussan) [inline verifiers/groth16/src/groth16.rs:52]: This should actually be
```rust
(crate::MAX_NUM_INPUTS + 1) * g1_size
```
You could add a test which builds a vk of maximum size, encodes it, and then checks that its actual size coincides with the one returned by `max_encoded_len()`

**Review by lgiussan — CHANGES_REQUESTED** [review]

File: verifiers/groth16/src/groth16.rs (hunk 5 above)
Comment (la10736) [reply verifiers/groth16/src/groth16.rs:52]: DOne.... and thx! I found a bug, ok minor because I been overestimating it but a bug !!!

**Review by lgiussan — APPROVED** [review]

**Review by lgiussan — COMMENTED** [review]
LGTM!
